raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -564,11 +640,31 @@ private FindQuorumResponseData handleFindQuorumRequest(FindQuorumRequestData req
             .setVoters(voters);
     }
 
-    private void handleFindQuorumResponse(FindQuorumResponseData response) throws IOException {
-        // We only need to become a follower if the current leader is not elected
+    private Optional<Errors> handleFindQuorumResponse(
+        RaftResponse.Inbound responseMetadata
+    ) throws IOException {
+        FindQuorumResponseData response = (FindQuorumResponseData) responseMetadata.data;
+        Errors error = Errors.forCode(response.errorCode());
+
+        // Regardless of error, we take the latest connection states from the response
+        updateVoterConnections(response);
+
         OptionalInt leaderIdOpt = optionalLeaderId(response.leaderId());
-        if (leaderIdOpt.isPresent() && leaderIdOpt.getAsInt() != quorum.localId) {
-            becomeFollower(leaderIdOpt.getAsInt(), response.leaderEpoch());
+        if (handleNonMatchingResponseLeaderAndEpoch(response.leaderEpoch(), leaderIdOpt)) {
+            // Unlike other API types, FindQuorum may indicate have a higher epoch or
+            // a different leaderId than our own and still return an error code of NONE,
+            // so we have to check for it
+            if (error == Errors.NONE)
```

Comment:
See my other comment: logically I think returning an `Optional.empty()` should have the same semantics as returning an `Optional.of(NONE)`.